core/chain_makers.go

```diff
@@ -358,32 +360,35 @@ func GenerateChainWithGenesis(genesis *Genesis, engine consensus.Engine, n int,
 		panic(err)
 	}
 	if genesis.Config != nil && genesis.Config.IsPrague(genesis.ToBlock().Number(), genesis.ToBlock().Time()) {
-		blocks, receipts, _, _ := GenerateVerkleChain(genesis.Config, genesis.ToBlock(), engine, db, n, gen)
+		blocks, receipts, _, _, _, _, _ := GenerateVerkleChain(genesis.Config, genesis.ToBlock(), engine, db, n, gen)
 		return db, blocks, receipts
 	}
 	blocks, receipts := GenerateChain(genesis.Config, genesis.ToBlock(), engine, db, n, gen)
 	return db, blocks, receipts
 }
 
-func GenerateVerkleChain(config *params.ChainConfig, parent *types.Block, engine consensus.Engine, db ethdb.Database, n int, gen func(int, *BlockGen)) ([]*types.Block, []types.Receipts, []*verkle.VerkleProof, []verkle.StateDiff) {
+func GenerateVerkleChain(config *params.ChainConfig, parent *types.Block, engine consensus.Engine, db ethdb.Database, n int, gen func(int, *BlockGen)) ([]*types.Block, []types.Receipts, []*verkle.VerkleProof, []verkle.StateDiff, [][][]byte, [][][]byte, [][][]byte) {
 	if config == nil {
 		config = params.TestChainConfig
 	}
 	proofs := make([]*verkle.VerkleProof, 0, n)
 	keyvals := make([]verkle.StateDiff, 0, n)
 	blocks, receipts := make(types.Blocks, n), make([]types.Receipts, n)
+	computedKeys := make([][][]byte, n)
+	computedPreStateValues := make([][][]byte, n)
+	computedPostStateValues := make([][][]byte, n)
 	chainreader := &generatedLinearChainReader{
 		config: config,
 		// GenerateVerkleChain should only be called with the genesis block
 		// as parent.
 		genesis: parent,
 		chain:   blocks,
 	}
-	genblock := func(i int, parent *types.Block, statedb *state.StateDB) (*types.Block, types.Receipts) {
+	genblock := func(i int, parent *types.Block, statedb *state.StateDB) (*types.Block, types.Receipts, [][]byte, [][]byte, [][]byte) {
```

Comment:
Now `genblock` will return some extra helpers that the test will use:
- `computedKeys`: the ordered keys collected in the witness during block execution.
- `computedPreStateValues`: the pre-state values from the tree.
- `computedPostStateValues`: we use the collected written values in the tree to calculate what we think are the correct post-state values.

The test will use these values to verify the proof in the block. 
To be clear: these three returned values are _trusted_ data since from the perspective of the client they come from his/her own computation. These will be compared against the block proof which, in theory, is _not_ trusted.

Now... it's true that in the case of these test, the block proof is generated by the same client that is verifying the proof. So, for the `computedKeys` and `computedPreStateValues` seems pretty clear that they should match. What's more interesting is `computedPostStateValues`. For this, the verifying is using the collected tree writes, and the prover is using the tree directly (i.e: in `FinalizeAndAssemble(...)`).

The reality is that we should also make the prover to use these collected values, so the proving calculation of post-state values is faster.

But I'd prefer to do that in another PR since, if I do that now, this test won't be actually "testing" much really since both the prover and verifier would be doing the same calculation for post state values.